### PR TITLE
fixed usage of Parameters with non-numeric key in nested attributes

### LINKED ISSUE #29888: Strong parameters doesn't permit all nested attributes if key is not numeric
### Steps to reproduce
ppp = ActionController::Parameters.new( {"user_infos_attributes"=>{"1500663412001"=>{"address"=>"222", "city"=>"222"}, 'NEW_RECORD' => {"address" => "333", "city" => "333"}}})
aaa = [{:user_infos_attributes => [:address, :city]}]
ppp.permit(aaa)

vs

With removed non-numeric key 'NEW_RECORD':
ppp = ActionController::Parameters.new( {"user_infos_attributes"=>{"1500663412001"=>{"address"=>"222", "city"=>"222"}}})
aaa = [{:user_infos_attributes => [:address, :city]}]
ppp.permit(aaa)

(Guidelines for creating a bug report are [available
here](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#creating-a-bug-report))

### Expected behavior
Tell us what should happen
ppp.permit(aaa) should give a correct set:
<ActionController::Parameters {"user_infos_attributes"=><ActionController::Parameters {"1500663412001"=><ActionController::Parameters {"address"=>"222", "city"=>"222"} permitted: true>} permitted: true>} permitted: true>

"Strong parameters" permitting should not think that keys of nested records are always numeric ('NEW_RECORD' or '12332435345'), anyway the keys are not used. Or at least only the record with non-numeric key should be unpermitted, but not all of the passed records.

### Actual behavior
Tell us what happens instead
ppp.permit(aaa) gives
<ActionController::Parameters {"user_infos_attributes"=><ActionController::Parameters {} permitted: true>} permitted: true>

### System configuration
**Rails version**:
Rails 5.0.4.
**Ruby version**:
2.4.1

### PR DESCRIPTION
**Pull request by lfekson**
### Summary
Fixes #29888
"Parameters" object that includes nested attributes as { "tests_attributes" => { "111" => { "name" => "aaa" }, "new_record" => { "name" => "bbb" } } } now doesn't remove the whole "tests_attributes" hash, but keeps  { "tests_attributes" => { "111" => { "name" => "aaa" } } }  part of it.

### CODE REVIEW

**Comment by dguettler** [comment]
@lfekson this should be added to the description
Fixes #29888

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 1]
```diff
@@ -216,6 +216,12 @@ class Parameters
     #    config.always_permitted_parameters = %w( controller action format )
     cattr_accessor :always_permitted_parameters, default: %w( controller action )
 
+    class << self
+      def field_style_valid?(k = nil, v = nil)
```
Comment (gmcgibbon) [inline actionpack/lib/action_controller/metal/strong_parameters.rb]: You probably don't need to use default nil arguments here, do you?

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 2]
```diff
@@ -800,7 +806,7 @@ def permitted=(new_permitted)
       end
 
       def fields_for_style?
-        @parameters.all? { |k, v| k =~ /\A-?\d+\z/ && (v.is_a?(Hash) || v.is_a?(Parameters)) }
+        @parameters.any? { |k, v| Parameters.field_style_valid?(k, v) }
```
Comment (gmcgibbon) [inline actionpack/lib/action_controller/metal/strong_parameters.rb]: This makes sense, we're only checking if one or more a numeric. Maybe we should rename this method (and the class level one too) to `#nested_attributes?` or `#nested?` instead. I think `#field_for_style?` is too married to the concept of form data.

File: actionpack/test/controller/parameters/nested_parameters_permit_test.rb [hunk 3]
```diff
@@ -150,6 +150,29 @@ def assert_filtered_out(params, key)
     assert_filtered_out permitted[:book][:authors_attributes]["0"], :age_of_death
   end
 
+  test "fields_for-style nested params with non_numeric keys" do
+    params = ActionController::Parameters.new(
+      book: {
+        authors_attributes: {
+          '0': { name: "William Shakespeare", age_of_death: "52" },
+          '1': { name: "Unattributed Assistant" },
+          'new_record': { name: "Some name" }
```
Comment (gmcgibbon) [inline actionpack/test/controller/parameters/nested_parameters_permit_test.rb]: You should add a `'2': "NOT A HASH"` nested attribute to this test to make sure that case gets filtered too.

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 4]
```diff
@@ -852,7 +858,7 @@ def each_element(object)
         when Parameters
           if object.fields_for_style?
             hash = object.class.new
-            object.each { |k, v| hash[k] = yield v }
+            object.each { |k, v| hash[k] = yield v if Parameters.field_style_valid?(k, v) }
```
Comment (gmcgibbon) [inline actionpack/lib/action_controller/metal/strong_parameters.rb]: I wonder if adding a `#each_nested`, `#each_field` or something like that that does the filtering for you is worth it.

**Review by gmcgibbon — COMMENTED** [review]
Still relevant in `6.0.0.alpha`. See comments for some suggestions 👍

**Comment by kamipo** [comment]
Closing in favor of #36127.

**Pull request closed without merging**